### PR TITLE
WDPD-209 Create Vault table

### PR DESCRIPTION
**Pull request by ps-cgrach**
### This PR

Create the `Vault` table to save credit card information.

### Jira Ticket
https://jira.parkside.at/browse/WDPD-209

### CODE REVIEW

File: Core/Vault.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+     * @since 1.3.0
+     */
+    public static function getCards()
+    {
+        $oUser = Registry::getSession()->getUser();
+        $sUserId = $oUser->getId();
+        $sAddressId = $oUser->getSelectedAddressId();
+        $aCards = [];
+        try {
+            $sQuery = "SELECT * from " . OxidEeEvents::VAULT_TABLE . " " .
```
Comment (ps-holler) [inline Core/Vault.php]: The string concatenation at the end is not necessary when using double quotes.
```suggestion
            $sQuery = "SELECT * from " . OxidEeEvents::VAULT_TABLE . " 
```

File: Core/Vault.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        $sMaskedPan = $oResponse->getMaskedAccountNumber();
+        $iExporationMonth = $aCard['expiration-month'];
+        $iExpirationYear = $aCard['expiration-year'];
+        $oUser = Registry::getSession()->getUser();
+        $sUserId = $oUser->getId();
+        $sAddressId = $oUser->getSelectedAddressId();
+
+        $aExistingCards = self::getCards($sUserId, $sAddressId);
+        foreach ($aExistingCards as $aCard) {
+            if ($aCard['TOKEN'] == $sToken && $aCard['ADDRESSID'] == $sAddressId) {
```
Comment (ps-holler) [inline Core/Vault.php]: It would be safer to also check for the type here, e.g. use `===` instead of `==`.

File: Core/Vault.php (hunk 1 above)
Comment (ps-cgrach) [reply Core/Vault.php]: 👍

File: Core/Vault.php (hunk 2 above)
Comment (ps-cgrach) [reply Core/Vault.php]: 👍

File: Core/Vault.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+     * @since 1.3.0
+     */
+    public static function getCards()
+    {
+        $oUser = Registry::getSession()->getUser();
+        $sUserId = $oUser->getId();
+        $sAddressId = $oUser->getSelectedAddressId();
+        $aCards = [];
+        try {
+            $sQuery = "SELECT * from " . OxidEeEvents::VAULT_TABLE . "
```
Comment (tomazpu) [inline Core/Vault.php]: Can we move this to a private getCardsFromDB the sql and try/catch block

File: Core/Vault.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        $sAddressId = $oUser->getSelectedAddressId();
+
+        $aExistingCards = self::getCards();
+        foreach ($aExistingCards as $aCard) {
+            if ($aCard['TOKEN'] === $sToken && $aCard['ADDRESSID'] === $sAddressId) {
+                return;
+            }
+        }
+
+        $sQuery = "INSERT INTO " . OxidEeEvents::VAULT_TABLE . " SET
```
Comment (tomazpu) [inline Core/Vault.php]: Same here a private method for the insert to the DB, even better would be to have a common DB class where just crud is done. Now i have the feeling that DB operations are all over the code. What do you think?

File: Core/Vault.php [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+     * @return int
+     *
+     * @throws DatabaseConnectionException
+     * @throws DatabaseErrorException
+     *
+     * @since 1.3.0
+     */
+    public static function deleteCard($sVaultId)
+    {
+        $sQuery = "DELETE FROM " . OxidEeEvents::VAULT_TABLE . "
```
Comment (tomazpu) [inline Core/Vault.php]: Same here

File: Core/Vault.php (hunk 4 above)
Comment (ps-cgrach) [reply Core/Vault.php]: Moving to a separate class does not make sense - we use the plain SQL queries only here. Moving it to a separate function sounds good.

File: Core/Vault.php (hunk 3 above)
Comment (ps-cgrach) [reply Core/Vault.php]: 👍

File: Core/Vault.php (hunk 4 above)
Comment (ps-cgrach) [reply Core/Vault.php]: Like discusses, moved to methods

File: Core/Vault.php (hunk 5 above)
Comment (ps-cgrach) [reply Core/Vault.php]: :+1

**Pull request merged**